templates/cluster-template.yaml

```diff
@@ -0,0 +1,222 @@
+          #!/bin/bash
+          export DEBIAN_FRONTEND=noninteractive
+          hostnamectl set-hostname "$1" && hostname -F /etc/hostname
+          mkdir -p -m 755 /etc/apt/keyrings
+          VERSION=${2%.*}
+          curl -fsSL "https://pkgs.k8s.io/core:/stable:/v$VERSION/deb/Release.key" | sudo gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
+          echo "deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/v$VERSION/deb/ /" | sudo tee /etc/apt/sources.list.d/kubernetes.list
+          apt-get update -y
+          apt-get install -y kubelet kubeadm kubectl containerd
+          apt-mark hold kubelet kubeadm kubectl containerd
+          modprobe overlay
+          modprobe br_netfilter
+          sysctl --system
+        permissions: "0777"
+    preKubeadmCommands:
+      - /kubeadm-init.sh '{{ ds.meta_data.label }}' "${KUBERNETES_VERSION}"
+    clusterConfiguration:
+      apiServer:
+        extraArgs:
+          cloud-provider: external
+        timeoutForControlPlane: 20m
+    initConfiguration:
+      nodeRegistration:
+        kubeletExtraArgs:
+          cloud-provider: external
+          provider-id: 'linode:///{{ ds.meta_data.region }}/{{ ds.meta_data.id }}'
+        name: '{{ ds.meta_data.label }}'
+    joinConfiguration:
+      nodeRegistration:
+        kubeletExtraArgs:
+          cloud-provider: external
+          provider-id: 'linode:///{{ ds.meta_data.region }}/{{ ds.meta_data.id }}'
+        name: '{{ ds.meta_data.label }}'
+  version: "${KUBERNETES_VERSION}"
+---
+kind: LinodeMachineTemplate
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
+metadata:
+  name: ${CLUSTER_NAME}-control-plane
+spec:
+  template:
+    spec:
+      image: ${LINODE_OS}
+      type: ${LINODE_CONTROL_PLANE_MACHINE_TYPE}
+      region: ${LINODE_REGION}
+      authorizedKeys:
+        - ${LINODE_SSH_KEY}
+---
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDeployment
+metadata:
+  name: ${CLUSTER_NAME}-md-0
+spec:
+  clusterName: ${CLUSTER_NAME}
+  replicas: ${WORKER_MACHINE_COUNT}
+  selector:
+    matchLabels:
+  template:
+    spec:
+      clusterName: ${CLUSTER_NAME}
+      version: "${KUBERNETES_VERSION}"
+      bootstrap:
+        configRef:
+          name: ${CLUSTER_NAME}-md-0
+          apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+          kind: KubeadmConfigTemplate
+      infrastructureRef:
+        name: ${CLUSTER_NAME}-md-0
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
+        kind: LinodeMachineTemplate
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
+kind: LinodeMachineTemplate
+metadata:
+  name: ${CLUSTER_NAME}-md-0
+spec:
+  template:
+    spec:
+      image: ${LINODE_OS}
+      type: ${LINODE_MACHINE_TYPE}
+      region: ${LINODE_REGION}
+      authorizedKeys:
+        - ${LINODE_SSH_KEY}
+---
+apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+kind: KubeadmConfigTemplate
+metadata:
+  name: ${CLUSTER_NAME}-md-0
+spec:
+  template:
```

Comment:
Is there a way we can have this template apply to both the controlplane and worker nodes?

Reply:
`KubeadmControlPlaneTemplate` (control plane nodes only) and `KubeadmConfigTemplate` (worker nodes only) don’t seem like they can inherit a common base template AFAICT? I can do some more poking around to see if we can't simplify this.

Reply:
Looks like we can put the common files into a k8s Secret and then reference that in `KubeadmConfigTemplate.spec.template.spec.files.contentFrom` and `KubeadmControlPlaneTemplate.spec.template.spec.kubeadmConfigSpec.files.con
tentFrom`. We'd have to use a Secret since it doesn't seem like there's ConfigMap support. I can make that tweak on this template if we're okay with that approach to deduplicate the data.